src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -7949,17 +7949,44 @@ private GotoStatementSyntax ParseGotoStatement()
         private IfStatementSyntax ParseIfStatement()
         {
             Debug.Assert(this.CurrentToken.Kind == SyntaxKind.IfKeyword || this.CurrentToken.Kind == SyntaxKind.ElseKeyword);
-
             bool firstTokenIsElse = this.CurrentToken.Kind == SyntaxKind.ElseKeyword;
             var @if = firstTokenIsElse
                 ? this.EatToken(SyntaxKind.IfKeyword, ErrorCode.ERR_ElseCannotStartStatement)
                 : this.EatToken(SyntaxKind.IfKeyword);
+
+            var exclamationToken = this.CurrentToken.Kind == SyntaxKind.ExclamationToken
```

Comment:
>this.CurrentToken.Kind == SyntaxKind.ExclamationToken [](start = 35, length = 53)

Consider changing parsing logic only if the ```ExclamationToken``` is followed by the ```OpenParenToken``` to keep the diagnostics and the parse tree as they were before this feature.

Reply:
Hi @AlekseyTs .  I'll look into that.  I think it will depend on which produces more sensible errors in the case where you write something like `if !non_paren_expr` (which people are likely to try).  Thanks!

Reply:
> I think it will depend on which produces more sensible errors in the case where you write something like if !non_paren_expr (which people are likely to try).  

I think, if user is targeting C# 7.3, we better not suggest (with diagnostics) to add parentheses where they wouldn't be allowed anyway for that language version.


---
In reply to: [240412458](https://github.com/dotnet/roslyn/pull/30945#discussion_r240412458) [](ancestors = 240412458)